File: solutions/2026-01-10/python/2026-01-10-minimum-ascii-delete-sum-for-two-strings.py
```python
# Problem Summary: Find the minimum ASCII sum of deleted characters to make two strings equal.
# Link: https://leetcode.com/problems/minimum-ascii-delete-sum-for-two-strings/
# Approach: This problem can be solved using dynamic programming. We can define a DP table where dp[i][j] represents the minimum ASCII delete sum to make the first i characters of s1 and the first j characters of s2 equal.
#
# The recurrence relation is as follows:
# 1. If s1[i-1] == s2[j-1]:
#    The characters match, so no deletion is needed for these characters. The cost is the same as making the first i-1 characters of s1 and the first j-1 characters of s2 equal.
#    dp[i][j] = dp[i-1][j-1]
#
# 2. If s1[i-1] != s2[j-1]:
#    We have two options:
#    a. Delete s1[i-1]: The cost is the ASCII value of s1[i-1] plus the minimum cost to make the first i-1 characters of s1 and the first j characters of s2 equal.
#       cost1 = ord(s1[i-1]) + dp[i-1][j]
#    b. Delete s2[j-1]: The cost is the ASCII value of s2[j-1] plus the minimum cost to make the first i characters of s1 and the first j-1 characters of s2 equal.
#       cost2 = ord(s2[j-1]) + dp[i][j-1]
#    We choose the option that results in the minimum cost.
#    dp[i][j] = min(cost1, cost2)
#
# Base cases:
# - dp[0][0] = 0 (empty strings are equal with no deletions)
# - dp[i][0] = sum of ASCII values of the first i characters of s1 (to make s1[:i] equal to an empty string)
# - dp[0][j] = sum of ASCII values of the first j characters of s2 (to make an empty string equal to s2[:j])
#
# The final answer will be dp[len(s1)][len(s2)].
#
# Time Complexity: O(m*n), where m is the length of s1 and n is the length of s2. We iterate through the DP table of size (m+1) x (n+1).
# Space Complexity: O(m*n) for the DP table. This can be optimized to O(min(m, n)) by using only two rows or one row of the DP table, as each cell only depends on the previous row/column. For simplicity and clarity, we'll use O(m*n) space.

class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        m, n = len(s1), len(s2)

        # Initialize DP table with dimensions (m+1) x (n+1)
        # dp[i][j] will store the minimum delete sum for s1[:i] and s2[:j]
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        # Base case: Fill the first row
        # To make an empty string equal to s2[:j], we need to delete all characters in s2[:j]
        for j in range(1, n + 1):
            dp[0][j] = dp[0][j - 1] + ord(s2[j - 1])

        # Base case: Fill the first column
        # To make s1[:i] equal to an empty string, we need to delete all characters in s1[:i]
        for i in range(1, m + 1):
            dp[i][0] = dp[i - 1][0] + ord(s1[i - 1])

        # Fill the rest of the DP table
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                # If the current characters match, no deletion is needed for these characters.
                # The cost is inherited from the previous subproblem dp[i-1][j-1].
                if s1[i - 1] == s2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    # If the characters don't match, we have two options:
                    # 1. Delete s1[i-1]: cost is ord(s1[i-1]) + dp[i-1][j] (cost to make s1[:i-1] and s2[:j] equal)
                    # 2. Delete s2[j-1]: cost is ord(s2[j-1]) + dp[i][j-1] (cost to make s1[:i] and s2[:j-1] equal)
                    # We choose the option with the minimum cost.
                    dp[i][j] = min(ord(s1[i - 1]) + dp[i - 1][j], ord(s2[j - 1]) + dp[i][j - 1])

        # The final answer is in dp[m][n], representing the minimum delete sum for the entire s1 and s2.
        return dp[m][n]

```

File: solutions/2026-01-10/cpp/2026-01-10-minimum-ascii-delete-sum-for-two-strings.cpp
```cpp
/*
Problem Summary: Find the minimum ASCII sum of deleted characters to make two strings equal.
Link: https://leetcode.com/problems/minimum-ascii-delete-sum-for-two-strings/

Approach:
This problem can be solved using dynamic programming, similar to the Longest Common Subsequence (LCS) problem.
We want to find the longest common subsequence (in terms of ASCII values) between the two strings.
The characters that are part of this common subsequence will NOT be deleted.
The characters that are NOT part of this common subsequence WILL be deleted.

Let dp[i][j] be the minimum ASCII delete sum to make s1[0...i-1] and s2[0...j-1] equal.

Base cases:
- dp[0][0] = 0 (empty strings are equal with no deletions)
- dp[i][0] = sum of ASCII values of s1[0...i-1] (to make s1[0...i-1] equal to an empty string, delete all characters from s1)
- dp[0][j] = sum of ASCII values of s2[0...j-1] (to make s2[0...j-1] equal to an empty string, delete all characters from s2)

Transitions:
1. If s1[i-1] == s2[j-1]:
   The last characters match. We don't need to delete them. The minimum delete sum is the same as the sum for s1[0...i-2] and s2[0...j-2].
   dp[i][j] = dp[i-1][j-1]

2. If s1[i-1] != s2[j-1]:
   The last characters do not match. We have three options:
   a. Delete s1[i-1]: The cost is ASCII of s1[i-1] + minimum delete sum for s1[0...i-2] and s2[0...j-1].
      Cost = s1[i-1] + dp[i-1][j]
   b. Delete s2[j-1]: The cost is ASCII of s2[j-1] + minimum delete sum for s1[0...i-1] and s2[0...j-2].
      Cost = s2[j-1] + dp[i][j-1]
   c. Delete both s1[i-1] and s2[j-1]: This case is implicitly covered by options a and b. If we delete both, it's like considering the problem for s1[0..i-2] and s2[0..j-2] and adding the ASCII of both characters. However, this won't be optimal compared to deleting just one if the other character could form a longer common subsequence.
   We take the minimum of the costs from deleting s1[i-1] or s2[j-1].
   dp[i][j] = min(s1[i-1] + dp[i-1][j], s2[j-1] + dp[i][j-1])

The final answer will be dp[s1.length()][s2.length()].

Alternatively, we can compute the total ASCII sum of both strings and subtract twice the ASCII sum of their Longest Common Subsequence (LCS). This approach might be conceptually simpler if we already have an LCS function. However, the DP approach directly calculates the minimum delete sum, which is what the problem asks for.

Let's stick to the direct DP approach described above.

Time Complexity:
O(m*n), where m is the length of s1 and n is the length of s2.
We iterate through the entire dp table of size (m+1) x (n+1).

Space Complexity:
O(m*n) for the dp table. This can be optimized to O(min(m,n)) space by observing that each row only depends on the previous row.
However, for clarity and ease of implementation, we will use O(m*n) space.
*/

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

class Solution {
public:
    int minimumDeleteSum(std::string s1, std::string s2) {
        int m = s1.length();
        int n = s2.length();

        // dp[i][j] will store the minimum ASCII delete sum to make
        // s1[0...i-1] and s2[0...j-1] equal.
        std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));

        // Initialize base cases:
        // If one string is empty, we must delete all characters from the other string.

        // dp[i][0]: Make s1[0...i-1] equal to an empty string.
        // This means deleting all characters from s1[0...i-1].
        for (int i = 1; i <= m; ++i) {
            dp[i][0] = dp[i - 1][0] + s1[i - 1];
        }

        // dp[0][j]: Make s2[0...j-1] equal to an empty string.
        // This means deleting all characters from s2[0...j-1].
        for (int j = 1; j <= n; ++j) {
            dp[0][j] = dp[0][j - 1] + s2[j - 1];
        }

        // Fill the DP table
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                // If the current characters match
                if (s1[i - 1] == s2[j - 1]) {
                    // We don't need to delete these characters.
                    // The minimum delete sum is the same as the sum for the prefixes excluding these characters.
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    // If the current characters do not match, we have two options:
                    // 1. Delete s1[i-1]: Cost is ASCII of s1[i-1] plus the minimum delete sum
                    //    for s1[0...i-2] and s2[0...j-1].
                    // 2. Delete s2[j-1]: Cost is ASCII of s2[j-1] plus the minimum delete sum
                    //    for s1[0...i-1] and s2[0...j-2].
                    // We take the minimum of these two options.
                    dp[i][j] = std::min(dp[i - 1][j] + s1[i - 1], dp[i][j - 1] + s2[j - 1]);
                }
            }
        }

        // The final answer is the minimum delete sum to make the entire s1 and s2 equal.
        return dp[m][n];
    }
};
```